### ISSUE TITLE
File Locking fails to make files readonly unless I make a new clone

### ISSUE THREAD

**Issue by oleg620** [git-lfs/git-lfs#3246]
We decided to take advantage of the [file locking capability](https://github.com/git-lfs/git-lfs/wiki/File-Locking) on an existing repo.

We set `lfs.setlockablereadonly=true` for certain extensions and we made `git config lfs.setlockablereadonly` return true.

I hoped I could make locking work on existing clones as we have a large number of developers (~100) working on this large repo (over 30 min clone time).
I deleted some of the files and did `git reset --hard` assuming that the files would be recreated as readonly, but that didn't happen.
The only way I know of to make this work is to create a new clone.

Can this be fixed? Or do you have a workaround?
Thanks.

**Comment by ttaylorr** [comment]
Hi, Oleg. Thanks for letting us know.

> We set `lfs.setlockablereadonly=true` for certain extensions and we
> made `git config lfs.setlockablereadonly` return true.

Just to double check, have you run `git lfs track <ext> --lockable`? The
two are different, but I think that the difference can sometimes be
confusing.

  * `git lfs track` with `--lockable` tells Git LFS that the filetype
    can be "locked".

  * `lfs.setlockablereadonly=true` tells Git LFS how to _interpret_ what
    "locked" means. In particular, when true, it tells Git LFS to make
    files that are lockable and not locked by the current user
    read-only, while files locked by the current user are writeable.

Since `lfs.setlockablereadonly` is set to 'true' by default, is it
possible that you haven't run `git lfs track --lockable` yet?

If you don't want to run `git lfs track --lockable`, you can
alternatively change lines in your .gitattributes from:

    *.txt filter=lfs diff=lfs merge=lfs -text

To:

    *.txt filter=lfs diff=lfs merge=lfs lockable -text

> The only way I know of to make this work is to create a new clone.

Do the files appear as you expect them to on a new clone? In this case,
you can run 'git checkout --force' to trigger Git LFS's post-checkout
hook (which you can ensure exists by running 'git lfs install').

**Comment by ttaylorr** [comment]
> > Since `lfs.setlockablereadonly` is set to 'true' by default
> It was explicitly set to false in our initial setup.

I think that this might be why Git LFS is refusing to mark files as
writable/read-only, because of code in a number of places in the Git
LFS codebase that explicitly check
config.Config.SetLockableFilesReadOnly() before proceeding.

Do you perhaps have GIT_LFS_SET_LOCKABLE_READONLY set in your
environment?

> > Do the files appear as you expect them to on a new clone?
> Yes, they do.

I'm curious if there is an asymmetry between 'git clone' and 'git lfs
clone' in this regard. Which are you using when you describe a "fresh
clone"?

**Comment by sarvanguru** [comment]
I have tried with lfs file locking mechanism. I able to lock the file using below command 
$ git lfs lock test.txt
$ git push
$ $ git lfs locks

and also my team mate can able to unlock using --force option using below command 
$git lfs unlock test.txt  --force
file has been unlocked. 
and this was perfect unlocked. 

But, I am unable to edit the file after unlocked. 

Please suggest.

**Comment by ttaylorr** [comment]
> But, I am unable to edit the file after unlocked.

This is likely because your client hasn't asked for the updated state of
locks. Git LFS does not do so automatically without you first initiating
an operation which contacts the remote.

For example, if you run first:

  $ git lfs locks

Git LFS will try and contact the remote and ask for new locks, which
will inform it that your client no longer holds its lock on the file
which was '--force' unlocked by your coworker. It is at that point that
your client will remove the write-able bit and the locking state will be
restored.

The encouragement here is that, in general, one should make sure to
'sync' the local lock state before editing a file that you *think* you
have the lock for.

**Comment by sarvanguru** [comment]
Hi @ttaylorr  we are planning to migrate from Subversion to GIT for our one of the project in current assignments. hence, 
could you please provide the steps to be followed migrate from SVN to GIT ? 
we are using below version:
Tortoise SVN - 1.12.2 
GIT : 2.17.1.2 

this will be helpful for your reply! 

let me know if you need any info. 

Thanks,
Saravanan

**Comment by lcnittl** [comment]
I observe a similar behavior. Yet and most interesting, just for 1 out of 2 files.
We have 2 `*.docx` files in our repository, one gets `read-only` attribute set to true, the other one to false. It is worth mentioning that checkout for the one file actively sets `read-only` to false.

Would you have an idea what could cause such an issue?

Tracking as lockable is set up (and would also not explain the different behavior for the 2 files):
```
> git lfs tracks

Listing tracked patterns
    *.docx [lockable] (.gitattributes)
    *.mnova [lockable] (.gitattributes)
Listing excluded patterns

>
```

No files are locked:
```
> git lfs locks

>
```

Git version: `2.23.0.windows.1`
Git-LFS version: `2.8.0` (`git-lfs/2.8.0 (GitHub; windows amd64; go 1.12.2; git 30af66bb)`)

**Comment by lcnittl** [comment]
> 
> But, I am unable to edit the file after unlocked.
> This is likely because your client hasn't asked for the updated state of locks. Git LFS does not do so automatically without you first initiating an operation which contacts the remote. For example, if you run first: $ git lfs locks Git LFS will try and contact the remote and ask for new locks, which will inform it that your client no longer holds its lock on the file which was '--force' unlocked by your coworker. It is at that point that your client will remove the write-able bit and the locking state will be restored. The encouragement here is that, in general, one should make sure to 'sync' the local lock state before editing a file that you *think* you have the lock for.

@ttaylorr If I get this right, `git lfs locks` should update the lock state of a file for Git LFS. Does this also automatically set the `read-only` bit on lockable files, or is there another operation required to do so?

**Comment by bk2204** [comment]
I don't believe it does set the `read-only` bit.  You'd need some variant of `git checkout` (e.g., `git checkout .`) to do that.

**Comment by lcnittl** [comment]
Thanks for the reply! Any idea what could be wrong if a `git checkout .` (or even checking out the specific file) does not set the `read-only` bit after `git lfs locks`? A `git lfs lock file.ext && git lfs unlock file.ext` works, but is a bit clumsy :)

**Comment by dgow** [comment]
you can run `git lfs locks --verify && git checkout .` to sync locks
